astropy/io/fits/card.py

```diff
@@ -321,7 +321,11 @@ def value(self, value):
                            np.bool_)):
             raise ValueError(f'Illegal value: {value!r}.')
 
-        if isinstance(value, float) and (np.isnan(value) or np.isinf(value)):
+        if (isinstance(value, float) or
+            isinstance(value, numbers.Real)) and (np.isnan(value) or
```

Comment:
`numbers.Real` includes `float` so you need only one `isinstance` check.

Reply:
Done. But there is one issue. It appears one test case is failling for a `long_int` negative value, raising an unexpected `TypeError`. `np.isnan()` is not supported for the value, it seems.

![2021-07-05](https://user-images.githubusercontent.com/47617351/124509693-9da79700-dd8f-11eb-9d2c-2e7207f15759.png)

 Maybe I should filter out `int` values for the validation? Tested it that way and all test cases worked. Working code looks like this:
```
if (isinstance(value, numbers.Real) and not
    isinstance(value, int)) and (np.isnan(value) or
                                 np.isinf(value)):
```

I'll be waiting for your approval.

Reply:
Oh right, `numbers.Real` also catches integer which is not what we want here. Then I guess the best is to go back to `isinstance(value, (float, np.float32))`, sorry for the wrong direction.